disk.go

```diff
@@ -0,0 +1,190 @@
+	return nil
+}
+
+/* Search K8s for PersistentVolumeClaims with the snapshot policy annotation */
+func (m *diskManager) getDisks() ([]diskInfo, error) {
+	disks := make([]diskInfo, 0)
+
+	// get persistent volume claims
+	pvcs, err := m.k8s.CoreV1().PersistentVolumeClaims("").List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Error retrieving persistent volume claims: %v\n", err)
+	}
+	for _, pvc := range pvcs.Items {
+		if policy, ok := pvc.Annotations[m.config.TargetAnnotation]; ok {
+			// retrieve associated persistent volume for each claim
+			pv, err := m.k8s.CoreV1().PersistentVolumes().Get(pvc.Spec.VolumeName, metav1.GetOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("Error retrieving persistent volume: %s, %v\n", pvc.Spec.VolumeName, err)
+			}
+			diskName := pv.Spec.GCEPersistentDisk.PDName
+			logs.Info.Printf("found PersistentVolume: %q with disk: %q", pvc.GetName(), diskName)
+			disk := diskInfo{
+				name:   diskName,
+				policy: policy,
+			}
+			disks = append(disks, disk)
+		}
+	}
+
+	return disks, nil
+}
+
+/* Add the configured resource policy to the target disk */
+func (m *diskManager) addPolicy(info diskInfo) error {
+	// TODO only perform this api call if policyName is different
+	policy, err := m.getPolicy(info.policy)
+	if err != nil {
+		return fmt.Errorf("Error retrieving snapshot policy %s for disk %s: %v\n", info.policy, info.name, err)
+	}
+
+	disk, regional, err := m.findDisk(info.name)
+	if err != nil {
+		return err
+	}
+
+	// Check to see if any policies are already attached
+	if len(disk.ResourcePolicies) > 1 {
+		return fmt.Errorf("Disk %s has more than one resource policy, did the GCP API change? %v\n", info.name, disk.ResourcePolicies)
+	}
+	if len(disk.ResourcePolicies) == 1 {
+		if disk.ResourcePolicies[0] == policy.SelfLink {
+			logs.Info.Printf("Policy %s is already attached to disk %s, nothing to do\n", info.policy, info.name)
+			return nil
+		} else {
+			return fmt.Errorf("Unexpected policy %s is already attached to disk %s, please detach it manually and re-run\n", disk.ResourcePolicies[0], info.name)
+		}
+	}
+
+	// Attach policy
+	err = nil
+	if regional {
+		err = m.addPolicyToRegionalDisk(info.name, policy)
+	} else {
+		err = m.addPolicyToZonalDisk(info.name, policy)
+	}
+	if err != nil {
+		return fmt.Errorf("Error adding snapshot policy %s to disk %s: %v\n", info.policy, info.name, err)
+	}
+
+	logs.Info.Printf("Added policy %s to disk %s\n", info.policy, info.name)
+	return nil
+}
+
+/* Retrieve a regional or zonal disk object via the GCP API.
+   Returns the disk, a boolean that is true if the disk is regional, and an error
+*/
+func (m *diskManager) findDisk(name string) (*compute.Disk, bool, error) {
+	disk, err1 := m.getZonalDisk(name)
+	if err1 == nil {
+		logs.Info.Printf("Found disk %s in zone %s\n", name, m.config.Zone)
+		return disk, false, nil
+	}
+
+	disk, err2 := m.getRegionalDisk(name)
```

Comment:
Good point! The reason I have two is I want to log both errors *only* if both calls fail, because the expected behavior is that only one will succeed... but I can definitely try and make this more graceful. (eg. Use a list call and iterate or something).

Reply:
Ah ok that makes sense, I'm good with it as is then 👍🏻